lib/ansible/modules/cloud/amazon/ec2_vol.py

```diff
@@ -373,117 +386,196 @@ def attach_volume(module, ec2, volume, instance):
+    return volume_dict, changed
 
-    # If we're just listing volumes then do nothing, else get the latest update for the volume
-    if state != 'list':
-        volume.update()
 
-    volume_info = {}
-    attachment = volume.attach_data
+def get_volume_info(volume):
 
+    attachment_data = get_attachment_data(volume)
     volume_info = {
-        'create_time': volume.create_time,
-        'encrypted': volume.encrypted,
-        'id': volume.id,
-        'iops': volume.iops,
-        'size': volume.size,
-        'snapshot_id': volume.snapshot_id,
-        'status': volume.status,
-        'type': volume.type,
-        'zone': volume.zone,
+        'create_time': volume.get('create_time'),
+        'encrypted': volume.get('encrypted'),
+        'id': volume.get('volume_id'),
+        'iops': volume.get('iops'),
+        'size': volume.get('size'),
+        'snapshot_id': volume.get('snapshot_id'),
+        'status': volume.get('state'),
+        'type': volume.get('volume_type'),
+        'zone': volume.get('availability_zone'),
         'attachment_set': {
-            'attach_time': attachment.attach_time,
-            'device': attachment.device,
-            'instance_id': attachment.instance_id,
-            'status': attachment.status
+            'attach_time': attachment_data.get('attach_time', None),
+            'device': attachment_data.get('device', None),
+            'instance_id': attachment_data.get('instance_id', None),
+            'status': attachment_data.get('state', None),
+            'deleteOnTermination': attachment_data.get('delete_on_termination', None)
         },
-        'tags': volume.tags
+        'tags': boto3_tag_list_to_ansible_dict(volume.get('tags'))
     }
-    if hasattr(attachment, 'deleteOnTermination'):
-        volume_info['attachment_set']['deleteOnTermination'] = attachment.deleteOnTermination
 
     return volume_info
 
 
+def get_mapped_block_device(instance_dict=None, device_name=None):
+    mapped_block_device = None
+    if not instance_dict:
+        return mapped_block_device
+    if not device_name:
+        return mapped_block_device
+
+    for device in instance_dict.get('block_device_mappings', []):
+        if device['device_name'] == device_name:
+            mapped_block_device = device
+            break
+
+    return mapped_block_device
+
+
+def ensure_tags(module, connection, res_id, res_type, tags, add_only):
+    changed = False
+
+    filters = ansible_dict_to_boto3_filter_list({'resource-id': res_id, 'resource-type': res_type})
+    cur_tags = None
+    try:
+        cur_tags = connection.describe_tags(Filters=filters)
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        module.fail_json_aws(e, msg="Couldn't describe tags")
+
+    purge_tags = bool(not add_only)
+    to_update, to_delete = compare_aws_tags(boto3_tag_list_to_ansible_dict(cur_tags.get('Tags')), tags, purge_tags)
```

Comment:
If cur_tags doesn't contain Tags the first argument will be None (which will make compare_aws_tags fail). If cur_tags always contains Tags then you can just use cur_tags['Tags'].